pkg/controllers/provisioning/provisioner.go

```diff
@@ -17,128 +17,148 @@ package provisioning
+	crmetrics "sigs.k8s.io/controller-runtime/pkg/metrics"
 )
 
 var (
-	MaxBatchDuration = time.Second * 10
-	MinBatchDuration = time.Second * 1
-	// MaxPodsPerBatch limits the number of pods we process at one time to avoid using too much memory
-	MaxPodsPerBatch = 2_000
+	defaultBatchOptions = BatchOptions{
+		IdleDuration: 1 * time.Second,
+		MaxDuration:  10 * time.Second,
+		MaxSize:      2_000,
+	}
 )
 
 // Provisioner waits for enqueued pods, batches them, creates capacity and binds the pods to the capacity.
 type Provisioner struct {
 	// State
 	*v1alpha5.Provisioner
-	pods    chan *v1.Pod
-	results chan error
-	done    <-chan struct{}
+	Batches sync.Map
+	Ctx     context.Context
 	Stop    context.CancelFunc
 	// Dependencies
 	cloudProvider cloudprovider.CloudProvider
 	kubeClient    client.Client
-	scheduler     *scheduling.Scheduler
-	launcher      *Launcher
-}
-
-func (p *Provisioner) Start(ctx context.Context) {
-	go func() {
-		logging.FromContext(ctx).Info("Starting provisioner")
-		for {
-			select {
-			case <-p.done:
-				logging.FromContext(ctx).Info("Stopping provisioner")
-				return
-			default:
-				if err := p.provision(ctx); err != nil {
-					logging.FromContext(ctx).Errorf("Provisioning failed, %s", err.Error())
-				}
-			}
-		}
-	}()
+	coreV1Client  corev1.CoreV1Interface
+	packer        *binpacking.Packer
 }
 
-func (p *Provisioner) provision(ctx context.Context) (err error) {
-	// Wait for a batch of pods
-	pods := p.Batch(ctx)
-	// Communicate the result of the provisioning loop to each of the pods.
-	defer func() {
-		for i := 0; i < len(pods); i++ {
-			select {
-			case p.results <- err: // Block until result is communicated
-			case <-p.done: // Leave if closed
-			}
-		}
-	}()
-	// Separate pods by scheduling constraints
-	schedules, err := p.scheduler.Solve(ctx, p.Provisioner, pods)
-	if err != nil {
-		return fmt.Errorf("solving scheduling constraints, %w", err)
-	}
-	// Get Instance Types, offering availability may vary over time
-	instanceTypes, err := p.cloudProvider.GetInstanceTypes(ctx, &p.Spec.Constraints)
+// Add (threadsafe) the pod to the provisioner and block until it's processed.
+// The caller is responsible for verifying that the pod was scheduled correctly.
+func (p *Provisioner) Add(ctx context.Context, pod *v1.Pod) error {
+	// Fail if constraints are not viable for pod
+	constraints, err := p.Spec.Constraints.Tighten(pod)
 	if err != nil {
-		return fmt.Errorf("getting instance types")
+		return err
 	}
-	// Launch capacity and bind pods
-	if err := p.launcher.Launch(ctx, p.Provisioner, schedules, instanceTypes); err != nil {
-		return fmt.Errorf("launching capacity, %w", err)
+
+	// Atomically create or use an existing batch.
+	value, ok := p.Batches.LoadOrStore(constraints.MustHash(), NewBatch(ctx, defaultBatchOptions))
+	batch := value.(*Batch)
+	if !ok {
+		go func() {
+			batch.Start()
+			p.provision(p.Ctx, constraints, batch)
+			batch.Stop()
+		}()
 	}
+	batch.Add(ctx, pod)
```

Comment:
If I understand this correctly, the reconcile loop will be blocked until the provisioning is finished. The controller specifics how many concurrent reconciles call (now it is 10).  Giving that we take about a minute to launch a node, does it mean that the maximum number of pods that it can process is around 10/minute? what will happen when we have 150 pods/second churn rate?

Reply:
We don't block the reconcile loop node creation. It ends as soon as the node is launched and the pod bound. The pod concurrency we use is 10,000.